[DX-SDK] Auth0 React quickstart (#9077)
auth0/docs: articles/quickstart/spa/auth0-react/01.md

:::

- `redirectUri`: The URL to where you'd like to redirect your users after they authenticate with Auth0. 

Once configured, `Auth0Provider` will store the authentication state of your users, the state of the SDK (is Auth0 ready to use?), and provide you with helper methods to log in and log out your users, which you can access using the `useAuth0()` hook.

:::panel Checkpoint
The SDK integration with React is complete. Run your app to verify that everything is working without any errors.
:::

## Add Login to Your Application

The Auth0 React SDK gives you tooling to easily create functionality to log in your users to your React application, such as a button or a link. After you have configured the `Auth0Provider`, you can create a login button as follows:

```javascript
import React from "react";
import { useAuth0 } from "@auth0/auth0-react";

const LoginButton = () => {
  const { loginWithRedirect } = useAuth0();

  return <button onClick={() => loginWithRedirect()}>Log In</button>;
};

export default LoginButton;
```

:::note

> This guide focuses on using the `useAuth0()` custom React Hook. If you are using class components, check out [these samples using the `withAuth0()` higher-order component] [link to class component samples].  

Author: Need a link to the class components sample here

---
title: Login
description: "Auth0 allows you to add authentication to your React application quickly and to gain access to user profile information. This guide demonstrates how to integrate Auth0 with any new or existing React application using the Auth0 React SDK."
budicon: 448
topics:
  - quickstarts
  - spa
  - react
  - login
github:
  path: 01-Login
contentType: tutorial
useCase: quickstart
---
<!-- markdownlint-disable MD002 MD034 MD041 -->

<%= include('../_includes/_getting_started', { library: 'React', callback: 'http://localhost:3000', returnTo: 'http://localhost:3000', webOriginUrl: 'http://localhost:3000', showLogoutInfo: true, showWebOriginInfo: true, new_js_sdk: true, show_install_info: false }) %>

> ## Step 1: Install the Auth0 React SDK

Author: Lets leave the step numbers out of the headings here.

Run the following command within your project directory to install the Auth0 React SDK:

```bash
npm install @auth0/auth0-react
```

> The SDK exposes methods and variables that help you integrate Auth0 with your React application organically using React Hooks or High-Order Components.

Author: I'm not sure "organically" is the right word here. Perhaps "idiomatically"?

Author: ```suggestion
The SDK exposes methods and variables that help you integrate Auth0 with your React application idiomatically using React Hooks or Higher-Order Components.
```

import React from "react";
import ReactDOM from "react-dom";
import App from "./App";
import { Auth0Provider } from "@auth0/auth0-react";

ReactDOM.render(
  <Auth0Provider
    domain="${account.namespace}"
    clientId="${account.clientId}"
    redirectUri={window.location.origin}
  >
    <App />
  </Auth0Provider>,
  document.getElementById("root")
);
```

The `Auth0Provider` component takes the following props:

- `domain` and `clientId`: The values of these properties correspond to the "Domain" and "Client ID" values present under the "Settings" of the single-page application that you registered with Auth0.

:::note
If you are using a [custom domain with Auth0](https://auth0.com/docs/custom-domains), the value of the domain prop is the value of your custom domain instead of the value reflected in the “Settings” tab.
:::

- `redirectUri`: The URL to where you'd like to redirect your users after they authenticate with Auth0. 

Once configured, `Auth0Provider` will store the authentication state of your users, the state of the SDK (is Auth0 ready to use?), and provide you with helper methods to log in and log out your users, which you can access using the `useAuth0()` hook.

> 

Author: I wonder if there's a task to end this section that could be something like "The SDK has now been installed into your application; try to run your app to verify that it still works as intended and nothing has broken".

> ## Step 3: Log In to Your App

Author: Let's use "Add Login to your Application" (and drop the step number)

The Auth0 React SDK gives you tooling to easily create functionality to log in your users to your React application, such as a button or a link. After you have configured the `Auth0Provider`, you can create a login button as follows:

```javascript
import React from "react";
import { useAuth0 } from "@auth0/auth0-react";

const LoginButton = () => {
  const { loginWithRedirect } = useAuth0();

  return <button onClick={() => loginWithRedirect()}>Log In</button>;
};

export default LoginButton;
```

:::note

> This guide focuses on using the `useAuth0()` custom React Hook. If you are using class components, check out [these samples using the `withAuth0()` high-order component] [link to class component samples].  

Author: ```suggestion
This guide focuses on using the `useAuth0()` custom React Hook. If you are using class components, check out [these samples using the `withAuth0()` higher-order component] [link to class component samples].  
```

Author: Remember to add the link here, I think it's missing.

:::

You can then use that button in any other component where you need the login functionality. You can conditionally render it by using the `isAuthenticated` property exposed by the `useAuth0()` hook as follows:

```javascript
import React from "react";
import { useAuth0 } from "@auth0/auth0-react";
import LoginButton from "./LoginButton";

function App() {
  const { isLoading, isAuthenticated } = useAuth0();

  return (
    <div className="App">
      {isLoading ? (
        <h2>Loading...</h2>
      ) : (
        <nav>{!isAuthenticated && <LoginButton />}</nav>
      )}
    </div>
  );
}

export default App;
```

> You first use the `isLoading` property to check if the SDK has been initialized and it's ready for use. You then use `isAuthenticated` to render the button conditionally: if Auth0 has not authenticated the user, render the login button.

Author: I'm not convinced yet that we need to show a React developer how to use a component they just created, we just want to show them how to initiate a login, which the previous snippet does.

I think if we want to show off `isLoading` and `isAuthenticated`, let's leave that for the later snippet that demonstrates getting the user profile like we discussed earlier.

I think otherwise we can be too prescriptive about how their button should be implemented into an app we have no idea about.

:::panel Checkpoint
Add the `LoginButton` button to your application. When you click it, verify that your React application redirects you to the [Auth0 Universal Login](https://auth0.com/universal-login) and that you can now log in or sign up using a simple username and password or a social provider. Once that's complete, Auth0 will redirect you back to your application using the value of the `redirectUri` that you used to configure the `Auth0Provider`.
:::

![Auth0 Universal Login](https://cdn.auth0.com/blog/universal-login/lightweight-login.png)

:::note
The Google social provider is enabled by default on new tenants. Auth0 offers you developer keys to test logging in with a [social identity provider](https://auth0.com/docs/connections/identity-providers-social). However, these developer keys have some limitations that may cause your app to behave differently. For more details on this and on how to fix it, read on [Test Social Connections with Auth0 Developer Keys](https://auth0.com/docs/connections/social/devkeys#limitations-of-developer-keys).
:::

> ## Step 4: Log Out of Your App

Author: Let's use "Add Logout to your Application" (and drop the step number)

Now that you can log in to your app, you also need a way to log out. You can create a logout button as follows:

```javascript
import React from "react";
import { useAuth0 } from "@auth0/auth0-react";

const LogoutButton = () => {
  const { logout } = useAuth0();

  return <button onClick={() => logout()}>Log Out</button>;
};

export default LogoutButton;
```

Then, you can render the `LoginButton` and `LogoutButton` conditionally by using the `isAuthenticated` property once again:

```javascript
import React from "react";
import { useAuth0 } from "@auth0/auth0-react";
import LoginButton from "./LoginButton";
import LogoutButton from "./LogoutButton";

function App() {
  const { isLoading, isAuthenticated } = useAuth0();

  return (
    <div className="App">
      {isLoading ? (
        <h2>Loading...</h2>
      ) : (
        <nav>{isAuthenticated ? <LogoutButton /> : <LoginButton />}</nav>
      )}
    </div>
  );
}

export default App;
```

> If Auth0 has authenticated the user, render the logout button; otherwise, render the login button.

Author: Similar comment as previous. Perhaps we don't need to show them how the button can be used, we just need to show how the hook is used to log out.

:::panel Checkpoint
Add the `LogoutButton` button to your application. When you click it, verify that your React application redirects you the address you specified as one of the "Allowed Logout URLs".
:::

> ## Step 5: Get User Profile Information

Author: Let's use "Show the User Profile" (and drop the step number)

The Auth0 React SDK helps you retrieve the profile information associated with logged-in users quickly in whatever component you need it, such as their name or profile picture, to personalize a page. The profile information is available through the `user` object exposed by the `useAuth0()` custom React Hook. Take this `Profile` component as an example:

```javascript
import React from "react";
import { useAuth0 } from "@auth0/auth0-react";

const Profile = () => {
  const { user } = useAuth0();

  return (
    <div>
      <img src={user.picture} alt={user.name} />
      <h2>{user.name}</h2>
      <p>{user.email}</p>

>     </div>

Author: If we wanted to demonstrate the use case of `isAuthenticated`, let's do it here and make sure we also link off to the docs on GitHub where more detail about that property (and others) can be found.

---
title: Login

> description: This tutorial demonstrates how to add user login to a React application using Auth0.

Reviewer: All examples in this quickstart use hooks. We also need to provide examples when using classes.

Reviewer 2: @sandrinodimattia @stevehobbsdev My proposed solution to this would be to have a QuickStart based on function components with Hooks and another based on class components and the `context` object. That's exactly what I plan to do for the blog posts. 

I could even go further and suggest that we need a dedicated QS that uses TypeScript -- it's seeing tremendous growth in the React community and production apps.

Author: I'm not sure a second quickstart is the answer here, but what would the quickstart doc look like if we did use TypeScript? It would look 99% the same, as many of the code samples in the doc would not need changing as their types can be inferred.

The actual sample, however, is currently being converted to use TypeScript by default.

> My proposed solution to this would be to have a QuickStart based on function components with Hooks and another based on class components and the context object.

We have docs and examples as to how our hook can be used with class components - we should just link to those and default to hooks for the quickstart. As you mentioned earlier, the nuances with classes vs. hooks could be explained in more detail in a deep-dive blog post.

I would in any case shy away from having two quickstarts.

Reviewer: ![image](https://user-images.githubusercontent.com/655409/83081363-cb94b500-a080-11ea-8521-bf3a570aee91.png)

Can we have a picker of some sorts? The quickstart will always be the same, but you should be able to choose between Hooks, Hooks with TS, Classes, Classes with TS

Author: @sandrinodimattia I agree this would be really awesome. [We do something like this on some of our docs](https://auth0.com/docs/api-auth/tutorials/authorization-code-grant-pkce) ([here is the source](https://github.com/auth0/docs/blob/master/articles/api-auth/tutorials/authorization-code-grant-pkce.md) as to how this is done).

@dan-auth0, can you explore implementing this for one of the snippets and see how you get on?

Reviewer 2: Addressed in CF.

> description: This tutorial demonstrates how to add user login to a React application using Auth0.

Reviewer: What if I'm using typescript? Will that be a separate quickstart?

Reviewer 2: 👍 Mentioned that here. I think it should be a separate quickstart.

https://github.com/auth0/docs/pull/9077/files#r431108321

Author: See my comment above - I think we should just have one quickstart but default to TypeScript. The samples in the doc likely won't change too much, but the _sample_ will be in TypeScript for reference.

Reviewer 2: Moved to CF.

## Install the SDK

Run the following command within your project directory to install the Auth0 React SDK:

```bash
npm install @auth0/auth0-react
```

> You may use `yarn` instead.

You can use the Auth0 React SDK to add authentication to your React app quickly, gain access to user profile information, and securely call APIs.

## Configure the SDK

Under the hood, the Auth0 React SDK uses [React Context](https://reactjs.org/docs/context.html) to manage the authentication state of your users and to expose helpful utility methods. One way to integrate Auth0 with your React app is to wrap your root component with an `Auth0Provider` that you can import from the SDK.

```javascript
import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';

import {Auth0Provider} from "@auth0/auth0-react";

ReactDOM.render(
  <Auth0Provider 
    domain="${account.namespace}"
    client_id="${account.clientId}"
    redirect_uri={window.location.origin}
  >
    <App />
  </Auth0Provider>,
  document.getElementById("root")
);
```

The `Auth0Provider` component takes three props:
 

> - `domain` and `client_id`: The values of these properties correspond to the "Domain" and "Client ID" values present under the "Settings" of the single-page application that you registered with Auth0.

Reviewer: Please also explain how to use your custom domain if configured. This will not show up in the "Domain" field.

Reviewer 2: Sandrino, I don't think this is something that has been covered in a quickstart before. Perhaps, is it something that rather belongs to the docs?

Reviewer: We need to add a link to docs then. Copying the "Domain" from the settings page is not something you do when using a custom domain.

- `redirect_uri`: The URL to where you'd like to redirect your users after they authenticate with Auth0. 

Once configured, `Auth0Provider` will store the authentication state of your users, the state of the Auth0 Client (is Auth0 ready to use?), and provide you with helper methods to log in and log out your users. You can access the content of the `Auth0Provider` with `useAuth0()`, which is a custom [React Hook](https://reactjs.org/docs/hooks-reference.html) exposed by the SDK.

## Add Login and Logout Buttons to Your App

The Auth0 React SDK makes it easy to create functionality to log in or log out your users from your React application, such as a button or link. After you have configured the `Auth0Provider`, you can use the `useAuth0()` custom React Hook within any component where you need this functionality by using the following boolean flags and methods it exposes:

- `isAuthenticated`: Returns `true` if the Auth0 authentication server has authenticated the user.

> - `isLoading`: Returns `true` if the SDK has created an Auth0 Client and has determined the user's authentication status.

Reviewer: If the user's authentication status **has been determined**, should the value not be false? Is "isLoading" not set to true when the user is still signing in?

Reviewer 2: True! this is a typo.

Reviewer 2: Yeah, this is a typo. This is why I think it's more helpful for this to be called something like `isReady` instead. It's better from a cognitive and development perspective to think in "positive", `true` terms. I feel it's better to check is Auth0 `isReady` rather than if it `isLoading`. 

Compare:

`isLoading`: Returns `false` when the SDK has created an Auth0 Client and has determined the user's authentication status.

vs:

`isReady`: Returns `true` when the SDK has created an Auth0 Client and has determined the user's authentication status.

As such, you can create an `Auth0ProviderWithHistory` function component that renders `Auth0Provider`. This wrapper component then calls `useHistory()` within its body to make the `history` object available within the scope of `Auth0Provider` and its  `onRedirectCallback` event handler.

**Root component:**

```javascript
ReactDOM.render(
  <BrowserRouter>
    <Auth0ProviderWithHistory>
      <App />
    </Auth0ProviderWithHistory>
  </BrowserRouter>,
  document.getElementById("root")
);
```

**`Auth0ProviderWithHistory` component:**

```javascript
const Auth0ProviderWithHistory = ({ children }) => {
  const history = useHistory();

  const onRedirectCallback = (appState) => {
    history.push(appState?.returnTo || window.location.pathname);
  };

  return (
    <Auth0Provider
      domain="${account.namespace}"
      client_id="${account.clientId}"
      redirect_uri={window.location.origin}

>       onRedirectCallback={onRedirectCallback}

Reviewer: Why does `onRedirectCallback` use came case with client_id and redirect_uri do not? We need consistency here.

Reviewer 2: I was wondering the same when I encountered it 🤔 It's also idiomatic to use camel case for variable names in JS/TS.

Author: To be fair this has come from the SPA SDK which has the same issue. I think the distinction is that `client_id` refers to a param on the query string when calling `/authorize`.

Not saying that's correct, that's just the way it is today. I would like to change this in the SPA SDK but will require a new major. We could address this now in the React SDK and map the params in the meantime.

cc: @adamjmcgrath

Reviewer 3: We can certainly rename the main ones (`client_id` etc). But while the SPA JS SDK let's you overload the constructor with any query params you want to send on the `/authorize` url - there will always be the possibility of mixing snake_case and camelCase.

We could do what the Express SDK does and use a container property like `authorizationParams` (https://github.com/auth0/express-openid-connect/blob/master/API.md#authorization-params-key) to pass extra stuff to `/authorize` I'll raise an issue to look into this.

    >
      {children}
    </Auth0Provider>
  );
};
```

The implementation of the authentication route guard may look like this `PrivateRoute` component:

```javascript
const PrivateRoute = ({ component, ...args }) => (

>   <Route component={withLoginRequired(component)} {...args} />

Reviewer: Where does withLoginRequired come from? No require statement.

Reviewer 2: 👍 I'll include the import.

);
```

You can use the `PrivateRoute` component in the same way that you use the `Route` component from React Router:

```javascript
function App() {
  const { isLoading } = useAuth0();

  return (
    <div className="App">
      {isLoading ? (
        <h2>Loading...</h2>
      ) : (
        <>
          <NavBar />
          <Switch>
            <Route path="/" exact component={Home} />
            <PrivateRoute path="/profile" component={Profile} />
          </Switch>
        </>
      )}
    </div>
  );
}
```

⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️

**TODO: Final URL to sample.**
 
⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️

Take a look at this sample that puts this technique together while also implementing a navigation bar: "Using React Router with the Auth0 React SDK" [sample required]

## Call an API

> Your React application may need to access protected data from an API that you are also [securing with Auth0](https://auth0.com/docs/microsites/protect-api/protect-api). For that use case, you'll need to retrieve an access token in the format of a [JSON Web Token (JWT)](https://jwt.io/) to authorize any calls you make to that secured API. 

Reviewer: This section doesn't say anything about scopes.

Reviewer 2: Sandrino, from your perspective, what should we cover about scopes in the QuickStart? One of the goals is to keep it short and quick as to demonstrate quickly the minimum integration of Auth0 with XYZ (in this case, React). We've felt that more nuanced and detailed concepts should be delegated to documentation and their practice to blog posts.

Reviewer: Ok let's add a link to some page that talks about scopes then. However, the SDK works in a specific way when it comes to scopes, we should also have a link to that.

Reviewer 2: Moved to CF.

> For the Auth0 React SDK to configure the access token correctly, you need to pass an `audience` property to the `Auth0Provider` as follows:

Reviewer: This page mentions nothing about the localhost issue & consent which you can bump into when using Auth0 from localhost. There should at least be some mention about this.

Reviewer 2: @sandrinodimattia @stevehobbsdev Is this considered a bug? Is this documented anywhere in the docs?

Reviewer: It's not a bug, it's by design. However, we need to provide customers with a path to solve it if they bump into the issue. A link to some page will help here. We want to avoid content like this showing up because we're not documenting this good enough: https://dev.to/kleeut/solving-consentrequired-auth0-spa-sdk-1le2

Reviewer 3: > https://dev.to/kleeut/solving-consentrequired-auth0-spa-sdk-1le2

This issue will be resolved by https://github.com/auth0/auth0-spa-js/pull/482 (the react sdk will use `checkSession` when it's released and this will allow users to handle recoverable errors like `consent_required` by having the user login again)

Reviewer 2: Moved to CF.

---
title: Login
description: This tutorial demonstrates how to add user login to a React application using Auth0.
budicon: 448
topics:
  - quickstarts
  - spa
  - react
  - login
github:
  path: 01-Login
contentType: tutorial
useCase: quickstart
---

<!-- markdownlint-disable MD002 MD034 MD041 -->

> This guide demonstrates how to use the Auth0 React SDK with any new or existing React application.

Reviewer 3: ```suggestion
This guide demonstrates how to use the Auth0 React SDK with any new or existing React applications.
```

## Prerequisites

- If you don't have an Auth0 account already, <a href="https://auth0.com/signup" data-amp-replace="CLIENT_ID" data-amp-addparams="anonId=CLIENT_ID(cid-scope-cookie-fallback-name)">sign up for a free Auth0 account</a>, and [create a tenant](https://auth0.com/docs/getting-started/create-tenant).

- [Register your React application](https://auth0.com/docs/dashboard/guides/applications/register-app-spa) as a single-page application with Auth0.

## Install the SDK

Run the following command within your project directory to install the Auth0 React SDK:

```bash
npm install @auth0/auth0-react
```

> > You may use `yarn` instead.

Author: This came up in some in some internal discussion last year during some research into quickstarts - the feeling was that this kind of thing was a bit redundant as, if you're a `yarn` user, then you already know how `yarn` can replace what `npm` is already doing. And if you're not already using `yarn`, then you're not likely to start using it just because we've mentioned it here.

Not sure what you think of that but just thought I'd point it out.

You can use the Auth0 React SDK to add authentication to your React app quickly, gain access to user profile information, and securely call APIs.

## Configure the SDK

Under the hood, the Auth0 React SDK uses [React Context](https://reactjs.org/docs/context.html) to manage the authentication state of your users and to expose helpful utility methods. One way to integrate Auth0 with your React app is to wrap your root component with an `Auth0Provider` that you can import from the SDK.

```javascript
import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';

> import {Auth0Provider} from "@auth0/auth0-react";

Reviewer 3: would be nice to have consistent quotes/double quotes for those like me with OCD about these things - I think CRA uses single quotes by default

Author: Spacing around `Auth0Provider`, too, as Prettier will format it this way.

ReactDOM.render(
  <Auth0Provider 
    domain="${account.namespace}"
    client_id="${account.clientId}"
    redirect_uri={window.location.origin}
  >
    <App />
  </Auth0Provider>,
  document.getElementById("root")
);
```

The `Auth0Provider` component takes three props:
 
- `domain` and `client_id`: The values of these properties correspond to the "Domain" and "Client ID" values present under the "Settings" of the single-page application that you registered with Auth0.

- `redirect_uri`: The URL to where you'd like to redirect your users after they authenticate with Auth0. 

Once configured, `Auth0Provider` will store the authentication state of your users, the state of the Auth0 Client (is Auth0 ready to use?), and provide you with helper methods to log in and log out your users. You can access the content of the `Auth0Provider` with `useAuth0()`, which is a custom [React Hook](https://reactjs.org/docs/hooks-reference.html) exposed by the SDK.

## Add Login and Logout Buttons to Your App

> The Auth0 React SDK makes it easy to create functionality to log in or log out your users from your React application, such as a button or link. After you have configured the `Auth0Provider`, you can use the `useAuth0()` custom React Hook within any component where you need this functionality by using the following boolean flags and methods it exposes:

Reviewer 3: "within any component" - this is where you could mention class vs functional components.

- `isAuthenticated`: Returns `true` if the Auth0 authentication server has authenticated the user.

- `isLoading`: Returns `false` when the SDK has created an Auth0 Client and has determined the user's authentication status.

> - `login()`: Takes your users to the [Auth0 Universal Login](https://auth0.com/docs/universal-login) page and back to your application once they log in.

Reviewer 3: FYI, am changing this to `loginWithRedirect` in `0.3.0`

  return (
    <div className="App">
      {isLoading ? (
        <h2>Loading...</h2>
      ) : (
        <>
          {isAuthenticated ? (
            <button onClick={() => logout()}>Log Out</button>
          ) : (
            <button onClick={() => login()}>Log In</button>
          )}
        </>
      )}
    </div>
  );
}

export default App;
```

The `App` component above displays a `Loading...` message while the SDK completes its initialization process: creating an Auth0 Client and determining the user's authentication status. Once the initialization is complete, the component renders a "Log In" button if Auth0 has not authenticated the user yet; otherwise, it renders a "Log Out" button.

> By default, Auth0 offers you developer keys to test logging in with a [social identity provider](https://auth0.com/docs/connections/identity-providers-social), such as Google. These developer keys have some limitations that may cause your app to behave differently. For more details on this and on how to fix it, read on ["Test Social Connections with Auth0 Developer Keys"](https://auth0.com/docs/connections/social/devkeys#limitations-of-developer-keys).

## Read the User Profile

> Once you have users stored in your Auth0 user database, the Auth0 React SDK helps you retrieve the profile information associated with logged-in users quickly in whatever component you need it, such as their name or profile picture to personalize a page. The profile information is available through the `user` object exposed by the `useAuth0()` custom React Hook.

Reviewer 3: Not sure you need "Once you have users stored in your Auth0 user database"

Take for example this `Profile` component:

```javascript
import React from "react";
import { useAuth0 } from "@auth0/auth0-react";

const Profile = () => {
  const { user } = useAuth0();

  return (
    <div>
      <img src={user.picture} alt={user.name} />
      <h2>{user.name}</h2>
      <p>{user.email}</p>
    </div>
  );
};

export default Profile;
```

> The `user` property from `useAuth()` contains sensitive information and artifacts related to the user's identity. As such, you can prevent leaking sensitive information to the public by checking if your app has initialized the SDK correctly and if Auth0 has authenticated the user before you render any component that consumes the `user` property. An alternative to doing manual checks within a component is to prevent the whole component from rendering using routing.

Reviewer 3: "As such, you can prevent leaking sensitive information to the public by checking if your app has initialized the SDK correctly" 

Don't think this is correct, you wouldn't leak that information if you didn't check the user was authenticated before visiting it, you'd just see a blank page.

Reviewer 2: Moved to CF.

⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️

**TODO: Final URL for additional routing examples.**
 
⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️

> React has no opinion on routing. This example assumes you are using [`react-router-dom`](https://github.com/ReactTraining/react-router#readme) &mdash; [access more examples for Gatsby and Next.js.](https://github.com/auth0-samples/auth0-react-sdk-sample)

The default behavior is to redirect your users to the value of the `redirect_uri` prop of `Auth0Provider` after they log in or sign up using Auth0. However, you may want to redirect users directly to the private route they were trying to access. You can define such behavior using the `onRedirectCallback` prop of `Auth0Provider`.

With React Router, your goal is to use the `history` object within the `onRedirectCallback` method to navigate to that private route programmatically. React Router exposes the `useHistory()` custom React Hook to make it easy to access the `history` object in any component that is a descendant of the `BrowserRouter` component.

> As such, you can create an `Auth0ProviderWithHistory` function component that renders `Auth0Provider`. This wrapper component then calls `useHistory()` within its body to make the `history` object available within the scope of `Auth0Provider` and its  `onRedirectCallback` event handler.

Reviewer 3: ```suggestion
As such, you can create an `Auth0ProviderWithHistory` function component that renders `Auth0Provider`. This wrapper component then calls `useHistory()` within its body to make the `history` object available within the scope of `Auth0Provider` and its `onRedirectCallback` event handler.
```

**Root component:**

```javascript
ReactDOM.render(
  <BrowserRouter>
    <Auth0ProviderWithHistory>
      <App />
    </Auth0ProviderWithHistory>
  </BrowserRouter>,
  document.getElementById("root")
);
```

**`Auth0ProviderWithHistory` component:**

```javascript
const Auth0ProviderWithHistory = ({ children }) => {
  const history = useHistory();

  const onRedirectCallback = (appState) => {
    history.push(appState?.returnTo || window.location.pathname);
  };

  return (
    <Auth0Provider
      domain="${account.namespace}"
      client_id="${account.clientId}"
      redirect_uri={window.location.origin}
      onRedirectCallback={onRedirectCallback}
    >
      {children}
    </Auth0Provider>
  );
};
```

The implementation of the authentication route guard may look like this `PrivateRoute` component:

```javascript

> import { withLoginRequired } from "@auth0/auth0-react";

Reviewer 3: FYI, this is going to be called `withAuthenticationRequired` in `0.3.0`

const PrivateRoute = ({ component, ...args }) => (
  <Route component={withLoginRequired(component)} {...args} />
);
```

You can use the `PrivateRoute` component in the same way that you use the `Route` component from React Router:

```javascript
function App() {
  const { isLoading } = useAuth0();

  return (
    <div className="App">

>       {isLoading ? (

Reviewer 3: You shouldn't need this loading check

Reviewer 2: Moved to CF.

            <Route path="/" exact component={Home} />
            <PrivateRoute path="/profile" component={Profile} />
          </Switch>
        </>
      )}
    </div>
  );
}
```

⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️

**TODO: Final URL to sample.**
 
⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️ ⚠️

Take a look at this sample that puts this technique together while also implementing a navigation bar: "Using React Router with the Auth0 React SDK" [sample required]

## Call an API

Your React application may need to access protected data from an API that you are also [securing with Auth0](https://auth0.com/docs/microsites/protect-api/protect-api). For that use case, you'll need to retrieve an access token in the format of a [JSON Web Token (JWT)](https://jwt.io/) to authorize any calls you make to that secured API. 

For the Auth0 React SDK to configure the access token correctly, you need to pass an `audience` property to the `Auth0Provider` as follows:

```javascript
import React from "react";
import ReactDOM from "react-dom";
import App from "./App";

import { Auth0Provider } from "@auth0/auth0-react";

ReactDOM.render(

>   <Auth0Provider

Reviewer 3: If the user was following this from start to finish, they might find it confusing that they have to undo the stuff they did in `Auth0ProviderWithHistory` or adapt this to match it.

Reviewer 2: Moved to CF.

    domain="${account.namespace}"
    client_id="${account.clientId}"
    audience="${apiIdentifier}"
    redirect_uri={window.location.origin}
  >
    <App />
  </Auth0Provider>,
  document.getElementById("root")
);
```

The set up is similar to the one discussed in the "Configure the SDK" section: you wrap your root component with `Auth0Provider` and use the "Settings" values of the single-page application you've registered with Auth0 to populate the values of `domain` and `client_id`. 

However, this time you also pass it an `audience` prop to allow the API to validate the access token properly. When you [set up an API](https://auth0.com/docs/getting-started/set-up-api) with Auth0, you define its audience value, which you can find as the "Identifier" value under the API "Settings."

Once you configure `Auth0Provider`, you can easily get the access token using the `getToken()` method from the `useAuth0()` custom React Hook wherever you need it. Take this `Account` component as an example:

```javascript
import React, { useEffect, useState } from "react";
import { useAuth0 } from "@auth0/auth0-react";

const Account = () => {

>   const { getToken } = useAuth0();

Reviewer 3: FYI this is going to be called `getAccessTokenSilently` in `0.3.0`

  const [data, setData] = useState(null);

  useEffect(() => {
    (async () => {
      const token = await getToken();

      const response = await fetch("http://localhost:7000/protected-hello", {
        headers: {

>           Authorization: `Bearer <%= "${token}" %>`,

Reviewer 3: ```suggestion
          Authorization: `Bearer ${token}`,
```

Author: @adamjmcgrath The escaping here is necessary for the article to play nice with doc's templating engine. They aren't rendered on the output.

        },
      });
      const json = await response.json();

      setData(json);
    })();
  }, [getToken]);

  return (

>     <div>{data ? <span>{data.message}</span> : <span>Loading...</span>}</div>

Reviewer 3: Error handling is probably too much for the quickstart - but I'd add a `loading` state in here

Reviewer 2: Moved to CF.
